File: 916-word-subsets/916-word-subsets.py
class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        
        def calculateFreqCount(s):
            freq = [0]*26
            for c in s:
                freq[ord(c)-ord("a")] += 1
            return freq
        
        res = []
        maxFreqAcrossB = [0]*26
        
        for word in words2:
            freq = calculateFreqCount(word)
            for i in range(0, 26):
                maxFreqAcrossB[i] = max(maxFreqAcrossB[i], freq[i])
        
        for word in words1:
            freq = calculateFreqCount(word)
            charCountAcrossA = True
            
            for i in range(0, 26):
                if freq[i] < maxFreqAcrossB[i]:
                    charCountAcrossA = False
                    break
            
            if charCountAcrossA:
                res.append(word)
        
        return res
    
        

File: 916-word-subsets/916-word-subsets.cpp
class Solution {
public:
    vector<int> calculateFreqCount(string& s){
        
        vector<int> freq(26, 0);
        for(auto c : s){
            freq[c-'a']++;
        }
        
        return freq;
    }
    
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        
        vector<string> res;
        vector<int> maxFreqCountAcrossB(26, 0);
        
        for(auto word : words2){
            vector<int> freqCount = calculateFreqCount(word);
            for(int i=0; i<26; i++){
                maxFreqCountAcrossB[i] = max(maxFreqCountAcrossB[i], freqCount[i]);
            }
        }
        
        for(auto word : words1){
            vector<int> freqCount = calculateFreqCount(word);
            bool foundChars = true;
            for(int i=0; i<26; i++){
                if(freqCount[i] < maxFreqCountAcrossB[i]){
                    foundChars = false;
                    break;
                }
            }
            
            if(foundChars) res.push_back(word);
        }
        
        
        return res;
    }
};